src/main/java/io/debezium/connector/vitess/VitessConnector.java

```diff
@@ -41,10 +46,49 @@ public Class<? extends Task> taskClass() {
 
     @Override
     public List<Map<String, String>> taskConfigs(int maxTasks) {
-        if (maxTasks > 1) {
-            throw new IllegalArgumentException("Only a single connector task may be started");
+        LOGGER.info("Calculating taskConfigs for {} tasks", maxTasks);
+        List<String> shards = null;
+        String storagePerTask = properties.get(VitessConnectorConfig.OFFSET_STORAGE_PER_TASK.name());
+        if (storagePerTask != null && storagePerTask.equalsIgnoreCase("true")) {
+            VitessConnectorConfig connectorConfig = new VitessConnectorConfig(Configuration.from(properties));
+            shards = getVitessShards(connectorConfig);
+        }
+        return taskConfigs(maxTasks, shards);
+    }
+
+    public List<Map<String, String>> taskConfigs(int maxTasks, List<String> shards) {
+        LOGGER.info("Calculating taskConfigs for {} tasks and shards: {}", maxTasks, shards);
+        String storagePerTask = properties.get(VitessConnectorConfig.OFFSET_STORAGE_PER_TASK.name());
+        if (storagePerTask != null && storagePerTask.equalsIgnoreCase("true")) {
+            int tasks = Math.min(maxTasks, shards.size());
```

Comment:
How do you detect if maxTasks changes?
From your PR description you said you are not handling in this PR.
My thinking is that this would keep things silent behind the scene; If users didn't fully understand the implication and think vitess connector can handle that, they will have data loss if changing the max.tasks. Similarly to the case if they toggle `vitess.offset.storage.per.task` from true to false.

I'm thinking if we should catch that to log a warning or add a flag to decide whether to error out or just warn on migration related scenarios.

Reply:
It's not easy to read the previous tasks.max value, by the time connector was called, the value in ConfigState has already updated.  We can try to detect the previous tasks.max by:
1. Fetch the keys in OffsetStorageReader, we don't have a list API in OffsetStorageReader so we need to guess read.  Currently the partition key in offsetStorageReader is in the form of task0_4, task1_4, so we should be able to try those keys to see whether the entry existed or not in offset storage.  If it doesn't, this is the first time connector is run with tasks.max=4.  This might still be fine if there is no previous run with other tasks.max value.  How do we know there is no previous offset history?  One way to blindly trying fetching partition keys: task0_1, task0_2, task0_3 for all the total_partition_number < 4, and left over partition keys in offset storage would generate a warning or error.  Of course this is assuming we would only scale up the number of tasks through the history
2. Have user explicitly specify the previous tasks.max

Reply:
For 1, it seems that the blindly reading old offsets approach definitely has some flows, such as that it couldn't tell whether the old offset was  the exactly the "previous generation". If you 

Sounds like specifying previous tasks.max will do. Can you elaborate how do you make this work? 
- Users have to remember the previous tasks.max all the time right?
- If they didn't specify previous tasks.max what would happen?

Another question I have is:
Does tasks.max=N guarantees N tasks will be created? From our previous discussions in the JIRA ticket I remember it was not the case, e.g. when number of shard, M < tasks.max, then only M tasks will be created, right? Will this become a blocker for future migration?

Reply:
For approach 1 to check previous partition keys in offset storage, the ideal way is to convince Kafka Connect people to add a iterator API in OffsetStorageReader, but that might take a couple of month to add that API and trickle back to this project.  In the interim, as a workaround we might just have to guess the previous partition key names by looping through task0_1, task0_2, task0_3 ..., this might get expensive for large number of partitions;

For approach 2: asking user to pass in previous task numbers.  If we want to task, we want to ask the previous number of tasks used (instead of previous tasks.max specified since the connector might create less number of tasks), the user can get the number of tasks used by querying on connector REST API before migration.

There is also approach 3: asking user to specify a generation number.  The partition key will take the form of taskId_generationNum.  if task keys for the current generation (e.g. 2) is not present (i.e. task0_2, task1_2, task2_2 ... not present) but we found task keys for the previous generation (i.e. task0_1) present, then this is a an error;

Or approach 4: ask user to specify a boolean argument: offsetExisted.  On connector startup, if we found no partitions for the current taskId_totalTasks but offsetExisted=true, it's an error;  Or reversely, if we found partition keys existed but offSetExisted=false it's also an error.   On first run of the connector, the user will specify tasks.max=initial_value, offsetExisted=false;  On the 2nd run, the user will need to update the param to: tasks.max=initial_value, offSetsExisted=true;   When the tasks.max is changing, the user will do the offset migration in kafka topic and update the task param to tasks.max=new_value, offSetsExisted=true;   If the user forgot to do the offset migration and start the connector with the above value, the connector will fail to start since it cannot find offset in the storage for that tasks.max.

Reply:
@shichao-an To move this issue forward, I think I will probably just proceed with approach 3: (a) has the partition key to take the form of <taskId>_<numTasks>_<genNum>, (b) and the user to specify the generation number for the current partitioning, genNum starts with 1 and will increase continuously for each new partitioning change (genNumber=0 is reserved for the current partitioning which only uses server key) (c) for the user who wants to handle offset automatic migration from previous task partitioning, she will specify previous numTasks param, so we should be able to find all the existing partition keys in offset storage using <taskId>_<prevNumTasks>_<genNum-1>;  if those previous partitioning keys doesn't exist, the program will error out;   (d) for the user who doesn't want do offset auto migration (e.g. some of the old keys doesn't exist anymore), she can specify offset_migration=false, so the program will basically take the current offset (the tail of the binlog queue) from Vitess ;  Although user needs to specify a partitioning generation number and a previousNumTasks, this approach is more precise during migration and it solves the offset auto migration (an obstacle I tried to avoid previously).

Reply:
The generation number approach sounds more robust to me. I wanted to bring it out since you mentioned it earlier, but somehow I forgot :) I will defer the implementation details to you.